packages/crypto/src/bls12_318/aggregate.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{errors::InvalidPoint, CryptoError};
+
+use super::points::{g1_from_fixed, g2_from_fixed, G1, G2};
+
+const G1_POINT_SIZE: usize = 48;
+const G2_POINT_SIZE: usize = 96;
+
+/// Takes a list of points in G1 (48 bytes each) and aggregates them.
+///
+/// This is like Aggregate from <https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-05>
+/// but works for signatures as well as public keys.
+pub fn bls12_381_aggregate_g1(points: &[u8]) -> Result<[u8; 48], CryptoError> {
+    if points.len() % G1_POINT_SIZE != 0 {
+        return Err(InvalidPoint::DecodingError {}.into());
+    }
+
+    let points_count = points.len() / G1_POINT_SIZE;
+
+    use rayon::prelude::*;
+
+    let points: Vec<[u8; 48]> = points
+        .chunks_exact(G1_POINT_SIZE)
+        .map(|data| {
+            let mut buf = [0u8; 48];
+            buf[..].clone_from_slice(data);
+            buf
+        })
+        .collect();
+
+    let mut decoded_points = Vec::with_capacity(points_count);
+    points
+        .par_iter()
+        .map(g1_from_fixed)
+        .collect_into_vec(&mut decoded_points);
+
+    let out: Result<Vec<G1>, CryptoError> = decoded_points.into_iter().collect();
+    let out = out?;
+
+    let out = g1_sum(&out);
+
+    Ok(out.to_compressed())
+}
+
+/// Takes a list of points in G2 (96 bytes each) and aggregates them.
+///
+/// This is like Aggregate from <https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-05>
+/// but works for signatures as well as public keys.
+pub fn bls12_381_aggregate_g2(points: &[u8]) -> Result<[u8; 96], CryptoError> {
+    if points.len() % G2_POINT_SIZE != 0 {
+        return Err(InvalidPoint::DecodingError {}.into());
+    }
+
+    let points_count = points.len() / G2_POINT_SIZE;
+
+    use rayon::prelude::*;
+
+    let points: Vec<[u8; 96]> = points
+        .chunks_exact(G2_POINT_SIZE)
+        .map(|data| {
+            let mut buf = [0u8; 96];
+            buf[..].clone_from_slice(data);
+            buf
+        })
+        .collect();
+
+    let mut decoded_points = Vec::with_capacity(points_count);
+    points
+        .par_iter()
+        .map(g2_from_fixed)
+        .collect_into_vec(&mut decoded_points);
+
+    let out: Result<Vec<G2>, CryptoError> = decoded_points.into_iter().collect();
+    let out = out?;
+
+    let out = g2_sum(&out);
+
+    Ok(out.to_compressed())
+}
+
+/// Creates a sum of points in G1.
+///
+/// This is fast since math is done on projective points. Parallelization does not help here
+/// for ~500 elements.
+#[inline]
+pub fn g1_sum(elements: &[G1]) -> G1 {
+    elements.iter().sum()
+}
+
+/// Creates a sum of points in G2.
+///
+/// This is fast since math is done on projective points. Parallelization does not help here
+/// for ~500 elements.
+#[inline]
+pub fn g2_sum(elements: &[G2]) -> G2 {
+    elements.iter().sum()
+}
```

Comment:
Assuming the whole point of `g1_sum`/`g2_sum` is unit-testing, can we make them non-pub?